// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: prodvana/common_config/task.proto

package common_config

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on TaskConfig with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TaskConfig) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TaskConfig with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in TaskConfigMultiError, or
// nil if none found.
func (m *TaskConfig) ValidateAll() error {
	return m.validate(true)
}

func (m *TaskConfig) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetProgram() == nil {
		err := TaskConfigValidationError{
			field:  "Program",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetProgram()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TaskConfigValidationError{
					field:  "Program",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TaskConfigValidationError{
					field:  "Program",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetProgram()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TaskConfigValidationError{
				field:  "Program",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetRetryConfig()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TaskConfigValidationError{
					field:  "RetryConfig",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TaskConfigValidationError{
					field:  "RetryConfig",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRetryConfig()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TaskConfigValidationError{
				field:  "RetryConfig",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return TaskConfigMultiError(errors)
	}

	return nil
}

// TaskConfigMultiError is an error wrapping multiple validation errors
// returned by TaskConfig.ValidateAll() if the designated constraints aren't met.
type TaskConfigMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TaskConfigMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TaskConfigMultiError) AllErrors() []error { return m }

// TaskConfigValidationError is the validation error returned by
// TaskConfig.Validate if the designated constraints aren't met.
type TaskConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TaskConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TaskConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TaskConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TaskConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TaskConfigValidationError) ErrorName() string { return "TaskConfigValidationError" }

// Error satisfies the builtin error interface
func (e TaskConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTaskConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TaskConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TaskConfigValidationError{}
